saleor/graphql/menu/mutations.py

```diff
@@ -196,6 +198,103 @@ def user_is_allowed(cls, user, input):
         return user.has_perm('menu.manage_menus')
 
 
+_MenuMoveOperation = namedtuple(
+    '_MenuMoveOperation', ('menu_item', 'parent', 'sort_order'))
+
+
+class MenuItemMove(BaseMutation):
+    menu_item = graphene.List(
+        MenuItem, description='Assigned menu to move within.')
+
+    class Arguments:
+        move = graphene.List(
+            MenuItemMoveInput,
+            required=True, description='The menu position data')
+
+    class Meta:
+        description = 'Moves items of menus'
+
+    @staticmethod
+    def pre_validate_move(move):
+        """Validate if the given move could be possibly possible."""
+        if move.parent_id:
+            if move.item_id == move.parent_id:
+                raise ValidationError({
+                    'parent': 'Cannot assign a node to itself.'})
+
+    @staticmethod
+    def post_validate_move(operation: _MenuMoveOperation):
+        """Validate if the given move will be actually possible."""
+
+        if operation.parent:
+            if operation.menu_item.is_ancestor_of(operation.parent):
+                raise ValidationError({
+                    'parent': (
+                        'Cannot assign a node as child of '
+                        'one of its descendants.')})
+
+    @classmethod
+    def move_to_operation(cls, info, move) -> _MenuMoveOperation:
+        menu_item = cls.get_node_or_error(
+            info, move.item_id,
+            field='item_id', only_type=MenuItem)
+        parent_node = None
+
+        if move.parent_id is not None:
+            parent_node = cls.get_node_or_error(
+                info, move.parent_id,
+                field='parent_id', only_type=MenuItem)
+
+        return _MenuMoveOperation(
+            menu_item=menu_item,
+            parent=parent_node,
+            sort_order=move.sort_order)
+
+    @classmethod
+    def clean_move(
+            cls, info, move_operations: List) -> List[_MenuMoveOperation]:
+
+        operations = []
+        for move in move_operations:
+            cls.pre_validate_move(move)
+            operation = cls.move_to_operation(info, move)
+            cls.post_validate_move(operation)
+            operations.append(operation)
+        return operations
+
+    @classmethod
+    def user_is_allowed(cls, user, input):
+        return user.has_perm('menu.manage_menus')
+
+    @staticmethod
+    def process_item(item: _MenuMoveOperation):
+        menu_item = item.menu_item  # type: models.MenuItem
+
+        # Move the parent if provided
+        if item.parent:
+            menu_item.move_to(item.parent)
+        # Remove the menu item's parent if was set to none (root node)
+        elif menu_item.parent_id:
+            menu_item.parent_id = None
+
+        # Move the menu item
+        if item.sort_order is not None:
+            menu_item.sort_order = item.sort_order
+
+        menu_item.save()
+
+    @classmethod
+    def perform_mutation(cls, root, info, move):
+        move = cls.clean_move(info, move)
```

Comment:
There is a lot of things going on in this mutation and I'm a little confused with switching between "moves" and "operations". I don't know how the others feel about it, but maybe some renaming would make it easier to understand. What do you think about:

```
operations = cls.clean_moves(moves)
```

and later:

```
for operation in operations:
    cls.perform_operation(operation)
    menu_items.append(operation.menu_item)
```

When looking at `process_items` I get confused what items are we actually operating on: menu items, moves or operations. This little renaming might help.